failure building with openssl@3, works with openssl@1.1

Building gitx I  got this error building the ObjectiveGit.framework.

```
Undefined symbols for architecture x86_64:
  "_EVP_PKEY_id", referenced from:
      __libssh2_ed25519_new_private_frommemory in libssh2.a(openssl.o)
      __libssh2_pub_priv_keyfile in libssh2.a(openssl.o)
      __libssh2_pub_priv_keyfilememory in libssh2.a(openssl.o)
ld: symbol(s) not found for architecture x86_64
```

Building succeeded adding LDFLAGS="-L/usr/local/opt/openssl@1.1/lib" to the ObjectiveGit.framework target build settings.

![image](https://user-images.githubusercontent.com/7188/153059617-b3d90aaf-16ca-4e52-a0c5-6d87494edb22.png)

see: https://github.com/gitx/gitx/issues/239#issuecomment-1032960138

What are your commands to build completely from the command line?

After clone I presume they start like this:

```
git submodule update --init --recursive
./script/bootstrap
```

Now getting this error:

```
Undefined symbols for architecture x86_64:
  "_EVP_PKEY_id", referenced from:
      __libssh2_ed25519_new_private_frommemory in libssh2.a(openssl.o)
```

More complete output from log:

```
Ld /Users/stephen/Library/Developer/Xcode/DerivedData/ObjectiveGitFramework-fyaphaesxegukpfgkwopelgpztry/Build/Products/Debug/ObjectiveGit.framework/Versions/A/ObjectiveGit normal (in target 'ObjectiveGit-Mac' from project 'ObjectiveGitFramework')
    cd /Users/stephen/dev/macos/objgit/objective-git
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -target x86_64-apple-macos10.9 -dynamiclib -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk -L/Users/stephen/Library/Developer/Xcode/DerivedData/ObjectiveGitFramework-fyaphaesxegukpfgkwopelgpztry/Build/Products/Debug -L. -LExternal -L/usr/local/opt/openssl/lib -L/Users/stephen/dev/macos/objgit/objective-git/External/build/lib -F/Users/stephen/Library/Developer/Xcode/DerivedData/ObjectiveGitFramework-fyaphaesxegukpfgkwopelgpztry/Build/Products/Debug -filelist /Users/stephen/Library/Developer/Xcode/DerivedData/ObjectiveGitFramework-fyaphaesxegukpfgkwopelgpztry/Build/Intermediates.noindex/ObjectiveGitFramework.build/Debug/ObjectiveGit-Mac.build/Objects-normal/x86_64/ObjectiveGit.LinkFileList -install_name @rpath/ObjectiveGit.framework/ObjectiveGit -Xlinker -rpath -Xlinker @loader_path/Frameworks -Xlinker -rpath -Xlinker @executable_path/../Frameworks -Xlinker -object_path_lto -Xlinker /Users/stephen/Library/Developer/Xcode/DerivedData/ObjectiveGitFramework-fyaphaesxegukpfgkwopelgpztry/Build/Intermediates.noindex/ObjectiveGitFramework.build/Debug/ObjectiveGit-Mac.build/Objects-normal/x86_64/ObjectiveGit_lto.o -Xlinker -export_dynamic -Xlinker -no_deduplicate -fobjc-arc -fobjc-link-runtime -fapplication-extension -lgit2 -force_load External/libgit2.a External/libssh2.a -lcrypto -lssl -lcurl -lz -liconv -lgit2 -framework Cocoa -compatibility_version 1 -current_version 1 -Xlinker -dependency_info -Xlinker /Users/stephen/Library/Developer/Xcode/DerivedData/ObjectiveGitFramework-fyaphaesxegukpfgkwopelgpztry/Build/Intermediates.noindex/ObjectiveGitFramework.build/Debug/ObjectiveGit-Mac.build/Objects-normal/x86_64/ObjectiveGit_dependency_info.dat -o /Users/stephen/Library/Developer/Xcode/DerivedData/ObjectiveGitFramework-fyaphaesxegukpfgkwopelgpztry/Build/Products/Debug/ObjectiveGit.framework/Versions/A/ObjectiveGit

ld: warning: dylib (/usr/local/opt/openssl/lib/libcrypto.dylib) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: dylib (/usr/local/opt/openssl/lib/libssl.dylib) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: linking against a dylib which is not safe for use in application extensions: /usr/local/opt/openssl/lib/libcrypto.dylib
ld: warning: linking against a dylib which is not safe for use in application extensions: /usr/local/opt/openssl/lib/libssl.dylib
ld: warning: object file (External/libssh2.a(agent.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(channel.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(global.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(hostkey.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(session.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(userauth.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(misc.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(openssl.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(kex.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(packet.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(transport.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(keepalive.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(comp.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(mac.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(pem.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(bcrypt_pbkdf.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(crypt.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(blowfish.o)) was built for newer macOS version (11.0) than being linked (10.9)
Undefined symbols for architecture x86_64:
  "_EVP_PKEY_id", referenced from:
      __libssh2_ed25519_new_private_frommemory in libssh2.a(openssl.o)
      __libssh2_pub_priv_keyfile in libssh2.a(openssl.o)
      __libssh2_pub_priv_keyfilememory in libssh2.a(openssl.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

```
$ find /usr/local -name "libssh*.a"
/usr/local/lib/libssh2.a
/usr/local/Cellar/libssh2/1.10.0/lib/libssh2.a

$ ls -l /usr/local/lib/libssh2.a
lrwxr-xr-x  1 stephen  admin  38 Sep 27 11:18 /usr/local/lib/libssh2.a -> ../Cellar/libssh2/1.10.0/lib/libssh2.a
```

In tests below I am now doing this to setup for building:

```
git clone git@github.com:gitx/objective-git.git
cd objective-git
git submodule update --init --recursive
script/bootstrap && script/update_libgit2
```

Then tried with and without `test` in the `xcodebuild` invocation:


```
xcodebuild -workspace ObjectiveGitFramework.xcworkspace -scheme "ObjectiveGit Mac" test archive ARCHS="x86_64"
```
Same error when building in xcode app:
```
Undefined symbols for architecture x86_64:
  "_EVP_PKEY_id", referenced from:
      __libssh2_ed25519_new_private_frommemory in libssh2.a(openssl.o)
      __libssh2_pub_priv_keyfile in libssh2.a(openssl.o)
      __libssh2_pub_priv_keyfilememory in libssh2.a(openssl.o)
ld: symbol(s) not found for architecture x86_64
```

```
xcodebuild -workspace ObjectiveGitFramework.xcworkspace -scheme "ObjectiveGit Mac" archive ARCHS="x86_64"
```
Different (but probably related??):
```
** ARCHIVE FAILED **


The following build commands failed:
	Ld /Users/stephen/Library/Developer/Xcode/DerivedData/ObjectiveGitFramework-fyaphaesxegukpfgkwopelgpztry/Build/Intermediates.noindex/ArchiveIntermediates/ObjectiveGit\ Mac/IntermediateBuildFilesPath/UninstalledProducts/macosx/ObjectiveGit.framework/Versions/A/ObjectiveGit normal (in target 'ObjectiveGit-Mac' from project 'ObjectiveGitFramework')
```

I see this as a probably major issue 
<img width="944" alt="image" src="https://user-images.githubusercontent.com/3314607/153722019-e7fe205b-56a0-4545-af2c-13dafbf01be2.png">

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: ../libgit2.a(auth_negotiate.c.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: ../libgit2.a(winhttp.c.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: ../libgit2.a(auth_negotiate.c.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: ../libgit2.a(winhttp.c.o) has no symbols
```

Which ends into this finally

```
ld: warning: object file (External/libssh2.a(crypt.o)) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: object file (External/libssh2.a(blowfish.o)) was built for newer macOS version (11.0) than being linked (10.9)
Undefined symbols for architecture x86_64:
  "_EVP_PKEY_id", referenced from:
      __libssh2_ed25519_new_private_frommemory in libssh2.a(openssl.o)
      __libssh2_pub_priv_keyfile in libssh2.a(openssl.o)
      __libssh2_pub_priv_keyfilememory in libssh2.a(openssl.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

The big question is: **Why do your objects have no symbols ?** 
It's only related to your local machine, because CI can build and link it

Thanks for all the help.

Do you have an `x86_64` mac? If so can you attach a log from your build that is successful. 

Does `./External/libgit2.a` in your build dir have symbols for `auth_negotiate` and `winhttp`?

```
$ nm -gC ./External/libgit2.a | grep auth_negotiate
./External/libgit2.a:auth_negotiate.c.o: no symbols
./External/libgit2.a(auth_negotiate.c.o):
./External/libgit2.a:winhttp.c.o: no symbols
```

I also noticed in the last `clang` invocation which generates the `Undefined symbols for architecture x86_64: "_EVP_PKEY_id"` error  a nonexistent directory is included.

```
ld: warning: directory not found for option '-L/Users/stephen/dev/macos/objgit/objective-git/External/build/lib'
```

--- a/ObjectiveGitFramework.xcodeproj/project.pbxproj
+++ b/ObjectiveGitFramework.xcodeproj/project.pbxproj
@@ -1687,7 +1687,7 @@
 				INFOPLIST_FILE = Info.plist;
 				LIBRARY_SEARCH_PATHS = (
 					"$(inherited)",
-					/usr/local/opt/openssl/lib,
+					"/usr/local/opt/openssl@1.1/lib",
 					"$(PROJECT_DIR)/External/build/lib",
 				);
 				MODULEMAP_FILE = ObjectiveGit.modulemap;
@@ -1716,7 +1716,7 @@
 				INFOPLIST_FILE = Info.plist;
 				LIBRARY_SEARCH_PATHS = (
 					"$(inherited)",
-					/usr/local/opt/openssl/lib,
+					"/usr/local/opt/openssl@1.1/lib",
 					"$(PROJECT_DIR)/External/build/lib",
 				);
 				MODULEMAP_FILE = ObjectiveGit.modulemap;
@@ -1956,7 +1956,7 @@
 				INFOPLIST_FILE = Info.plist;
 				LIBRARY_SEARCH_PATHS = (
 					"$(inherited)",
-					/usr/local/opt/openssl/lib,
+					"/usr/local/opt/openssl@1.1/lib",
 					"$(PROJECT_DIR)/External/build/lib",
 				);
 				MODULEMAP_FILE = ObjectiveGit.modulemap;
@@ -2176,7 +2176,7 @@
 				INFOPLIST_FILE = Info.plist;
 				LIBRARY_SEARCH_PATHS = (
 					"$(inherited)",
-					/usr/local/opt/openssl/lib,
+					"/usr/local/opt/openssl@1.1/lib",
 					"$(PROJECT_DIR)/External/build/lib",
 				);
 				MODULEMAP_FILE = ObjectiveGit.modulemap;

